_posts/2018-10-23-artsy-auth.md

```diff
@@ -0,0 +1,279 @@
+against our fooapp? We're going to create it manually and store it in barapp's environment. Let's go back to our
+Gravity console session - make sure `ca` is still defined. If you closed the session then before running the next
+code you should get a reference to that client application. `ca = ClientApplication.where(name: 'fooapp').first`
+should do it. Then:
+
+```
+  gravity:staging> expires_in = 10.years.from_now
+  gravity:staging> token = ApplicationTrust.create_for_token_authentication(ca, expires_in: expires_in)
+```
+
+This should generate a long string that it dumps to the console. Congratulations - that string is an encrypted JWT.
+Any http request with the header `Authorization: Bearer <encrypted>` will now be accepted by fooapp.
+
+So to use this JWT, we copy the string into barapp's environment - often with a second environment variable for the
+specific host we want to connect to. So in barapp you'd have environment variables like `FOO_JWT=<string>` and
+`FOO_HOST=http://fooapp.artsy.net` or `FOO_HOST=http://fooapp-staging.artsy.net` or whatever. (We store the host as
+an environment variable rather than having it in the code precisely so that we can target specific instances of the
+application depending on the desired context).
+
+Somewhere in barapp we'll have a service that sends HTTP requests to fooapp - we just have to make sure that those
+requests have the authorization header with the correct encrypted token, and bam! We're authenticated!
+
+So now it's time to put it all together - we know how to generate tokens, we know how to attach them to requests as
+authorization headers, and we know how to decrypt them locally in fooapp. So now what? What do we do with those
+tokens once we have them? That's where a useful library called ArtsyAuth comes in, at least if you're writing a
+Ruby on Rails application here at Artsy.
+
+### ArtsyAuth
+
+We've seen now how OAuth uses secret-key encryption to allow one service to vouch for a user to another service.
+But that's not quite enough - in our application, it's nice that we now have user data that we trust, but what do
+we do with it? How do we configure our application to allow trusted users? Come to think of it, where do we
+actually redirect unauthenticated users back to gravity to log in? OAuth is historically complex to set up
+correctly, but fortunately ArtsyAuth is here to help. It abstracts out most of the OAuth complexity for us,
+allowing us to focus on writing logic for what to do based on the JWT's actual values.
+
+A simple ruby gem that handles all of this for you. Just `bundle add artsy-auth` and then create
+`config/initializers/artsy_auth.rb`. It should look something like this:
+
+```
+  ArtsyAuth.configure do |config|
+    config.artsy_api_url = Osmosis.config.gravity_url # required
+    config.application_id = Osmosis.config.app_id # required
+    config.application_secret = Osmosis.config.app_secret # required
+  end
+```
+
```

Comment:
these larger code blocks could probably benefit from adding syntax highlighting by amending the opening of the code blocks to  ```ruby